skbio/tree/_bme.py

```diff
@@ -0,0 +1,352 @@
+        for j in range(i + 1, n):  # part (a)
+            b = ordered[j]
+            # skipping over ancestors
+            if b in ancestors:
+                continue
+            # both nodes are tips
+            if a.is_tip() and b.is_tip():
+                adm[i, j] = adm[j, i] = dm[a.name, b.name]
+            # second node is a tip, but not the first node
+            elif b.is_tip():
+                a1, a2 = a.children
+                for k, x in enumerate(ordered):
+                    if x is a1:
+                        i1 = k
+                    if x is a2:
+                        i2 = k
+                adm[i, j] = adm[j, i] = (adm[j, i1] + adm[j, i2]) * 0.5
+            # neither node is a tip
+            else:
+                b1, b2 = b.children
+                for k, x in enumerate(ordered):
+                    if x is b1:
+                        j1 = k
+                    if x is b2:
+                        j2 = k
+                adm[i, j] = adm[j, i] = (adm[j1, i] + adm[j2, i]) * 0.5
+    # calculating for second nodes which are ancestors
+    for j, b in enumerate(ordered):
+        (s,) = b.siblings()
+        p = b.parent
+        for k, x in enumerate(ordered):
+            if x is s:
+                j_s = k
+            if x is p:
+                j_p = k
+        for i, a in enumerate(ordered):
+            if b in a.ancestors():
+                adm[i, j] = adm[j, i] = (adm[i, j_s] + adm[i, j_p]) * 0.5
+    return adm
+
+
+def _bal_ols_edge(tree, dm):
+    """Assign estimated edge values to a tree based on a given distance matrix.
+
+    Estimation of edge values is based on an ordinary least squares (OLS) framework.
+    Average distances defined here coincide with the balanced minimum evolution problem.
+
+    """
+    adm = _balanced_average_matrix(tree, dm)
+    ordered = list(tree.postorder(include_self=False))
+    root = tree.root()
+    # identify edges by first finding the child node of an edge
+    for edge_node in ordered:
+        parent = edge_node.parent
+        # skip over root node
+        if edge_node.is_root():
+            continue
+        # calculate edge length for the edge adjacent to the root
+        elif parent.is_root():
+            for index, node in enumerate(ordered):
+                if node is edge_node:
+                    i1 = index
+            a, b = edge_node.children
+            for index, node in enumerate(ordered):
+                if node is a:
+                    i2 = index
+                elif node is b:
+                    i3 = index
+            edge_node.length = 0.5 * (adm[i2, i1] + adm[i3, i1] - adm[i2, i3])
+        # calculate edge lengths for external edges
+        elif edge_node.is_tip():
+            a = parent.parent
+            if a.is_root():
+                for child in a.children:
+                    a = child
+            b = edge_node.siblings()[0]
+            for index, node in enumerate(ordered):
```

Comment:
This `index` may be `i` because the word "index" may refer to row names in a dataframe. Whereas `i` is the typical way of denoting an index. Same below.